### ISSUE TITLE
map_unwrap_or incorrectly suggested due to ownership issues

### ISSUE THREAD

**Issue by demoray** [rust-lang/rust-clippy#10579]
### Summary

clippy is suggesting the use of `map_or` instead of `map(..).unwrap_or(..)` where the value of unwrap_or borrowed for the context of the map.

### Lint Name

clippy::map_unwrap_or

### Reproducer

I tried this code:

```rust
#[deny(clippy::map_unwrap_or)]
fn main() {
    let x = vec![1,2,3,0];
    let y = x.strip_suffix(&[0]).map(|s| s.to_vec()).unwrap_or(x);
    println!("{y:?}");
}
```

I saw this happen:

```
error: called `map(<f>).unwrap_or(<a>)` on an `Option` value. This can be done more directly by calling `map_or(<a>, <f>)` instead
 --> src/main.rs:4:13
  |
4 |     let y = x.strip_suffix(&[0]).map(|s| s.to_vec()).unwrap_or(x);
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_unwrap_or
note: the lint level is defined here
 --> src/main.rs:1:8
  |
1 | #[deny(clippy::map_unwrap_or)]
  |        ^^^^^^^^^^^^^^^^^^^^^
help: use `map_or(<a>, <f>)` instead
  |
4 -     let y = x.strip_suffix(&[0]).map(|s| s.to_vec()).unwrap_or(x);
4 +     let y = x.strip_suffix(&[0]).map_or(x, |s| s.to_vec());
  |

```

I expected to see this happen:
Not providing the warning.

Applying the suggestion results in the following:
```
error[[E0505]](https://doc.rust-lang.org/stable/error_codes/E0505.html): cannot move out of `x` because it is borrowed
 --> src/main.rs:5:41
  |
5 |     let y = x.strip_suffix(&[0]).map_or(x, |s| s.to_vec());
  |             -------------------- ------ ^ move out of `x` occurs here
  |             |                    |
  |             |                    borrow later used by call
  |             borrow of `x` occurs here

```

### Version

```text
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-unknown-linux-gnu
release: 1.68.2
LLVM version: 15.0.6
```
